README.md
EC-CUBE ショーケースプラグイン
==============================

EC-CUBE に商品を陳列する機能を追加します。


インストール方法
----------------

プラグイン設定
--------------

ラインセス
---------
同梱の LICENSE をご参照下さい。


連絡先
------
不具合を発見した場合や要望等がありましたら https://github.com/zenith6/eccube-showcase/issues までご連絡をお願いいたします。


作者
----
新田誠治 <zenith6@gmail.com> http://zenith6.github.io/


CHANGES.txt
## 1.0.0 (2014-**-**)

* 最初のリリース


tests/bootstrap.php
<?php

$loader = require_once dirname(__FILE__) . '/../src/vendor/autoload.php';


src/config.php
<?php
/*
 * ショーケースプラグイン
 * Copyright (C) 2014 Seiji Nitta All Rights Reserved.
 * http://zenith6.github.io/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

require_once PLUGIN_UPLOAD_REALDIR . 'Showcase/plugin_bootstrap.php';

$page = new plg_Showcase_Pages_ConfigPage();
$page->init();
$page->process();


src/Showcase.php
<?php
/*
 * ショーケースプラグイン
 * Copyright (C) 2014 Seiji Nitta All Rights Reserved.
 * http://zenith6.github.io/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/**
 * ショーケースプラグイン
 *
 * @author Seiji Nitta
 */
class Showcase extends SC_Plugin_Base
{
    /**
     * コンストラクタ
     * @param array $plugin_info プラグイン情報
     */
    public function __construct(array $plugin_info)
    {
        parent::__construct($plugin_info);
    }

    /**
     * プラグインをインストールします。
     *
     * @param array $plugin_info プラグイン情報
     * @param SC_Plugin_Installer $installer プラグインインストーラー
     * @return void
     */
    public function install(array $plugin_info, SC_Plugin_Installer $installer = null)
    {
        parent::install($plugin_info, $installer);

        $installer->copyFile('logo.png', 'logo.png');
    }

    /**
     * プラグインをアンインストールします。
     *
     * @param array $plugin_info プラグイン情報
     * @param SC_Plugin_Installer $installer プラグインインストーラー
     * @return void
     */
    public function uninstall(array $plugin_info, SC_Plugin_Installer $installer = null)
    {
        parent::uninstall($plugin_info, $installer);
    }

    /**
     * プラグインを有効化します。
     *
     * @param array $plugin_info プラグイン情報
     * @param SC_Plugin_Installer $installer プラグインインストーラー
     * @return void
     */
    public function enable(array $plugin_info, SC_Plugin_Installer $installer = null)
    {
        parent::enable($plugin_info, $installer);
    }

    /**
     * プラグインを無効化します。
     *
     * @param array $plugin_info プラグイン情報
     * @param SC_Plugin_Installer $installer プラグインインストーラー
     * @return void
     */
    public function disable(array $plugin_info, SC_Plugin_Installer $installer = null)
    {
        parent::disable($plugin_info, $installer);
    }

    /**
     * フックを登録します。
     *
     * @param SC_Helper_Plugin $pluginHelper プラグインヘルパー
     * @param int $priority
     */
    public function register(SC_Helper_Plugin $plugin_helper, $priority)
    {
        parent::register($plugin_helper, $priority);
    }
}


src/plugin_info.php
<?php
/*
 * ショーケースプラグイン
 * Copyright (C) 2014 Seiji Nitta All Rights Reserved.
 * http://zenith6.github.io/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/**
 * プラグインの情報を表します。
 *
 * @author Seiji Nitta
 */
class plugin_info
{
    public static $PLUGIN_CODE = 'Showcase';
    public static $PLUGIN_NAME = 'ショーケースプラグイン';
    public static $PLUGIN_VERSION = '1.0.0';
    public static $PLUGIN_SITE_URL = 'http://zenith6.github.com/eccube-showcase/';
    public static $DESCRIPTION = '商品の陳列機能を追加します。';
    public static $AUTHOR = 'Seiji Nitta';
    public static $AUTHOR_SITE_URL = 'http://zenith6.github.com/';
    public static $LICENSE = 'LGPL';
    public static $CLASS_NAME = 'Showcase';
    public static $COMPLIANT_VERSION = '2.13.1';
    public static $HOOK_POINTS = array();
}


src/plugin_update.php
<?php
/*
 * ショーケースプラグイン
 * Copyright (C) 2014 Seiji Nitta All Rights Reserved.
 * http://zenith6.github.io/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/**
 * プラグインのアップデートクラス
 *
 * @author Seiji Nitta
 */
class plugin_update
{
    /**
     * プラグインをアップデートします。
     *
     * @param array $plugin_info プラグイン情報
     * @param SC_Plugin_Installer $installer プラグインインストーラー
     * @return void
     */
    public function update(array $plugin_info, SC_Plugin_Installer $installer)
    {
    }
}


tasks/eccube/EccubeGetPluginInfoTask.php
<?php

/**
 * @author Seiji Nitta
 */
class EccubeGetPluginInfoTask extends Task
{

    /**
     * @var PhingFile
     */
    protected $pluginDir;

    /**
     * @var string
     */
    protected $key;

    /**
     * @var string
     */
    protected $outputProperty;

    public function main()
    {
        $plugin_dir = $this->pluginDir->getAbsolutePath();
        $info_file = $plugin_dir . '/plugin_info.php';
        if (!file_exists($info_file)) {
            throw new BuildException("Unable to read plugin_info.php file at " . $info_file);
        }

        require_once $info_file;
        $class = new ReflectionClass('plugin_info');
        $value = $class->getStaticPropertyValue($this->key);

        if ($this->outputProperty != '') {
            $this->project->setProperty($this->outputProperty, $value);
        } else {
            echo $value;
        }
    }

    public function setPluginDir(PhingFile $dir)
    {
        $this->pluginDir = $dir;
    }

    public function setKey($key)
    {
        $this->key = $key;
    }

    public function setOutputProperty($outputProperty)
    {
        $this->outputProperty = $outputProperty;
    }
}


src/library/plg/Showcase/Pages/ConfigPage.php
<?php
/*
 * ショーケースプラグイン
 * Copyright (C) 2014 Seiji Nitta All Rights Reserved.
 * http://zenith6.github.io/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

require_once CLASS_EX_REALDIR . 'page_extends/admin/LC_Page_Admin_Ex.php';

/**
 * プラグインの設定画面
 *
 * @author Seiji Nitta
 */
class plg_Showcase_Pages_ConfigPage extends LC_Page_Admin_Ex
{
    /**
     * @var Zeclib_PageStorage
     */
    public $storage;

    /**
     * @var array
     */
    public $errors = array();

    public function init()
    {
        parent::init();

        $this->template = PLUGIN_UPLOAD_REALDIR . 'Showcase/templates/admin/ownersstore/plg_Showcase_config.tpl';
        $this->tpl_subtitle = 'プラグイン設定';
    }

    public function process()
    {
        parent::process();

        $this->storage = $this->restoreStorage();

        $mode = $this->getMode();
        switch ($mode) {
            case 'save':
                $this->doSave();
                break;

            case 'edit':
            default:
                $this->doEdit();
                break;
        }
    }

    /**
     * @return Zeclib_PageStorage
     */
    protected function restoreStorage()
    {
        $storage = $this->createDefaultStorage();

        if (isset($_REQUEST['storage']) && is_string($_REQUEST['storage'])) {
            $encoded = $_REQUEST['storage'];
            $storage->restore($encoded);
        }

        return $storage;
    }

    /**
     * @return Zeclib_PageStorage
     */
    protected function createDefaultStorage()
    {
        $storage = new Zeclib_PageStorage(array(), AUTH_MAGIC);

        return $storage;
    }

    /**
     * @param SC_FormParam_Ex $params
     */
    protected function doEdit(SC_FormParam_Ex $params = null)
    {
        if (!$params) {
            $params = $this->buildFormParam();
            $this->setDefaultFormValues($params);
        }

        $this->form = $this->buildForm($params, $this->errors);
        $this->sendResponse();
    }

    /**
     * @param SC_FormParam_Ex $params
     */
    protected function setDefaultFormValues(SC_FormParam_Ex $params)
    {
    }

    /**
     * @throws Exception
     */
    protected function doSave()
    {
        try {
            $query = SC_Query_Ex::getSingletonInstance();
            $query->begin();

            $params = $this->buildFormParam();
            $params->setParam($_POST);

            $this->errors = $this->validateFormParams($params);
            if ($this->errors) {
                $query->rollback();
                $this->doEdit($params);
                return;
            }

            $this->applyFormParams($params);

            $query->commit();

            $this->tpl_javascript = "$(window).load(function () { alert('登録しました。'); });";
            $this->doEdit($params);
        } catch (Exception $e) {
            $query->rollback();

            throw $e;
        }
    }

    /**
     * @param SC_FormParam_Ex $params
     * @param array $errors
     * @return array
     */
    protected function buildForm(SC_FormParam_Ex $params, array $errors = array())
    {
        $form = array();

        foreach ($params->keyname as $index => $key) {
            $form[$key] = array(
                'title'     => $params->disp_name[$index],
                'value'     => $params->getValue($key),
                'maxlength' => $params->length[$index],
                'error'     => null,
            );
        }

        foreach ($errors as $key => $error) {
            $form[$key]['error'] = $error;
        }

        return $form;
    }

    /**
     * @return SC_FormParam_Ex
     */
    protected function buildFormParam()
    {
        $params = new SC_FormParam_Ex();

        return $params;
    }

    /**
     * @param SC_FormParam_Ex $params
     * @return array
     */
    protected function validateFormParams(SC_FormParam_Ex $params)
    {
        $errors = $params->checkError();

        return $errors;
    }

    /**
     * @param SC_FormParam_Ex $params
     */
    protected function applyFormParams(SC_FormParam_Ex $params)
    {
    }
}


src/plugin_bootstrap.php
<?php
/*
 * ショーケースプラグイン
 * Copyright (C) 2014 Seiji Nitta All Rights Reserved.
 * http://zenith6.github.io/
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

$loader = require_once dirname(__FILE__) . '/vendor/autoload.php';

$loader->add('plg_Showcase_Pages_', dirname(__FILE__) . '/library');
